File: server/services/indicators.py
import numpy as np
from models.stock import IndicatorValue


def calc_sma(closes: list[float], period: int) -> float | None:
    if len(closes) < period:
        return None
    return round(float(np.mean(closes[-period:])), 4)


def calc_ema(closes: list[float], period: int) -> list[float]:
    if not closes:
        return []
    k = 2 / (period + 1)
    result = [closes[0]]
    for i in range(1, len(closes)):
        result.append(closes[i] * k + result[-1] * (1 - k))
    return result


def calc_rsi(closes: list[float], period: int = 14) -> float | None:
    if len(closes) < period + 1:
        return None

    gains = []
    losses = []
    for i in range(1, len(closes)):
        diff = closes[i] - closes[i - 1]
        gains.append(max(diff, 0))
        losses.append(max(-diff, 0))

    avg_gain = np.mean(gains[:period])
    avg_loss = np.mean(losses[:period])

    for i in range(period, len(gains)):
        avg_gain = (avg_gain * (period - 1) + gains[i]) / period
        avg_loss = (avg_loss * (period - 1) + losses[i]) / period

    if avg_loss == 0:
        return 100.0
    rs = avg_gain / avg_loss
    return round(100 - (100 / (1 + rs)), 2)


def calc_macd(closes: list[float]) -> tuple[float, float, float] | None:
    if len(closes) < 35:  # need enough data for EMA26 + signal9
        return None
    ema12 = calc_ema(closes, 12)
    ema26 = calc_ema(closes, 26)
    macd_line = [a - b for a, b in zip(ema12, ema26)]
    signal_line = calc_ema(macd_line, 9)
    if not signal_line:
        return None
    val = round(macd_line[-1], 4)
    sig = round(signal_line[-1], 4)
    hist = round(val - sig, 4)
    return val, sig, hist


def calc_bollinger(closes: list[float], period: int = 20, mult: float = 2.0) -> tuple[float, float, float] | None:
    if len(closes) < period:
        return None
    window = closes[-period:]
    middle = float(np.mean(window))
    std = float(np.std(window))
    return (
        round(middle + mult * std, 4),
        round(middle, 4),
        round(middle - mult * std, 4),
    )


def compute_indicators(closes: list[float]) -> IndicatorValue:
    rsi_val = calc_rsi(closes)
    macd_result = calc_macd(closes)
    bb = calc_bollinger(closes)

    return IndicatorValue(
        rsi14=rsi_val,
        macd_value=macd_result[0] if macd_result else None,
        macd_signal=macd_result[1] if macd_result else None,
        macd_histogram=macd_result[2] if macd_result else None,
        bb_upper=bb[0] if bb else None,
        bb_middle=bb[1] if bb else None,
        bb_lower=bb[2] if bb else None,
        sma20=calc_sma(closes, 20),
        sma50=calc_sma(closes, 50),
        sma200=calc_sma(closes, 200),
    )

File: api/_lib/indicators.ts
/** Pure-math technical indicator calculations (no numpy dependency). */

export function calcSMA(closes: number[], period: number): number | null {
  if (closes.length < period) return null
  const slice = closes.slice(-period)
  return Math.round((slice.reduce((a, b) => a + b, 0) / period) * 10000) / 10000
}

function calcEMA(closes: number[], period: number): number[] {
  if (!closes.length) return []
  const k = 2 / (period + 1)
  const result = [closes[0]]
  for (let i = 1; i < closes.length; i++) {
    result.push(closes[i] * k + result[i - 1] * (1 - k))
  }
  return result
}

export function calcRSI(closes: number[], period = 14): number | null {
  if (closes.length < period + 1) return null
  const gains: number[] = []
  const losses: number[] = []
  for (let i = 1; i < closes.length; i++) {
    const diff = closes[i] - closes[i - 1]
    gains.push(Math.max(diff, 0))
    losses.push(Math.max(-diff, 0))
  }
  let avgGain = gains.slice(0, period).reduce((a, b) => a + b, 0) / period
  let avgLoss = losses.slice(0, period).reduce((a, b) => a + b, 0) / period
  for (let i = period; i < gains.length; i++) {
    avgGain = (avgGain * (period - 1) + gains[i]) / period
    avgLoss = (avgLoss * (period - 1) + losses[i]) / period
  }
  if (avgLoss === 0) return 100
  const rs = avgGain / avgLoss
  return Math.round((100 - 100 / (1 + rs)) * 100) / 100
}

export function calcMACD(
  closes: number[],
): { value: number; signal: number; histogram: number } | null {
  if (closes.length < 35) return null
  const ema12 = calcEMA(closes, 12)
  const ema26 = calcEMA(closes, 26)
  const macdLine = ema12.map((v, i) => v - ema26[i])
  const signalLine = calcEMA(macdLine, 9)
  if (!signalLine.length) return null
  const value = Math.round(macdLine[macdLine.length - 1] * 10000) / 10000
  const signal = Math.round(signalLine[signalLine.length - 1] * 10000) / 10000
  return {
    value,
    signal,
    histogram: Math.round((value - signal) * 10000) / 10000,
  }
}

export function calcBollinger(
  closes: number[],
  period = 20,
  mult = 2,
): { upper: number; middle: number; lower: number } | null {
  if (closes.length < period) return null
  const window = closes.slice(-period)
  const mean = window.reduce((a, b) => a + b, 0) / period
  const variance = window.reduce((sum, v) => sum + (v - mean) ** 2, 0) / period
  const std = Math.sqrt(variance)
  return {
    upper: Math.round((mean + mult * std) * 10000) / 10000,
    middle: Math.round(mean * 10000) / 10000,
    lower: Math.round((mean - mult * std) * 10000) / 10000,
  }
}

export function computeIndicators(closes: number[]) {
  const rsi = calcRSI(closes)
  const macd = calcMACD(closes)
  const bb = calcBollinger(closes)
  return {
    rsi14: rsi,
    macd_value: macd?.value ?? null,
    macd_signal: macd?.signal ?? null,
    macd_histogram: macd?.histogram ?? null,
    bb_upper: bb?.upper ?? null,
    bb_middle: bb?.middle ?? null,
    bb_lower: bb?.lower ?? null,
    sma20: calcSMA(closes, 20),
    sma50: calcSMA(closes, 50),
    sma200: calcSMA(closes, 200),
  }
}
